### DISCUSSION TITLE
Organization of various states in the program

### DISCUSSION THREAD

**Discussion by aadityadhruv** [hecrj/iced#2050] [Q&A]
I'm currently building an application to learn iced and advance my knowledge of rust. Right now, my program is structured such that I have a root struct which stores the states of everything, as such:

```rs

pub struct Process {
...//Some other stuff
currentScreen : ScreenStruct //See below
state : StateStruct,
... //more stuff

}
```
Currently I have 3 "screens" which I switch the `ScreenStruct` field in Process, and the view changes based on that field since I have a `match` statement for that. 

I also have a bunch of text input boxes in one of the screens, and for each of those text inputs I need to add an `Message` enum and a state to store the value of the text input. 

This is making my state struct grow large. I also don't know if the `match` statement in the `view` method is the correct way to handle different screens. 

Any advice on solving these issues would be appreciated!

**Comment by CelebrateVC** [comment]
Hi @aadityadhruv !  

I'm coming here for a similar issue, going off the doc's example of a mulit-view app, the `tour` it has General navigation messages in 1 Enum with a variant dedicated to "StepMessages", defined later
https://github.com/iced-rs/iced/blob/817f72868746461891ca4e74473c555f3b5c5703/examples/tour/src/main.rs#L99-L104
...
https://github.com/iced-rs/iced/blob/817f72868746461891ca4e74473c555f3b5c5703/examples/tour/src/main.rs#L213-L228
and then a pile of `if let STEP1 = self` nested inside of a match 
https://github.com/iced-rs/iced/blob/817f72868746461891ca4e74473c555f3b5c5703/examples/tour/src/main.rs#L231-L243 
This approach (or better yet making a new enum for each state, injected as a variant of the `StepMessages` which is itself in `Message`) is sustainable for medium scale projects.

From what I have been able to make out there is no way of making [Invalid States Unrepresentable](https://geeklaunch.io/blog/make-invalid-states-unrepresentable/#state-machines-and-state-transitions) with StateMachines and Transitions (because `trait Application` has `fn update(&mut self, message: Self::Message) -> Command<Self::Message>` it doesn't let you transition between different states of a state machine over a generic `<T>` and so you have to cover for all states+messages even if they are not valid. An explicit catch-all panic!() sweeps those under the rug to become a runtime error if they ever come up

It would be nice to have a way to define messages per state, but that's not able to be done at the moment. Below is a snip of my current match on my pet project.
![image](https://github.com/iced-rs/iced/assets/27698263/727b3f77-d60e-4145-990d-e4c3004d5827)
